### ISSUE TITLE
runtime::Printer has a `scratch` field that is ~never used

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#6469]
The runtime Printer has a `scratch` field that (presumably) is meant to avoid using `malloc()` for small print jobs, but in practice, it appears to ~never be used: 
- The length of `scratch` is (theoretically) a template parameter, which defaults to 1024
- ...but `scratch` is actually declared as `char scratch[length <= 256 ? length : 1];`, so any length of > 256 becomes just 1 (which in practice means "always use malloc")
- The places in runtime that override the default are:
    - profiler.cpp (which happens to also end up passing 1024)
    - tracing.cpp (which passes 4096)
    - opencl.cpp (16384)
    - d3d12compute.cpp (64k)
    - d3d12compute.cpp (16)
    - d3d12compute.cpp (256)

So basically, there are only two places (in d3d12compute) that are ever using this field.

The whole thing looks wonky to me -- my guess is that length was a straight passthrough in the past, but got limited this way due to stack-size issues on some architecture, and the ramifications never thought thru.

Anyway, this is an ugly wart that should really be resolved somehow, either by removing scratch entirely (since it's rarely used currently), or by deciding that a 1024-byte sized stack scratch is OK after all, or by deciding that  256-byte scratch is adequate for most things, or... something else.

(Note that d3d12compute.cpp has a `StackPrinter` class that attempts to make use of this, but the class is unused and should be removed.)

**Comment by shoaibkamil** [comment]
Good catch, yikes!  I definitely did not understand this was the behavior.

Pinging @slomp to chime in about the one in `d3d12compute.cpp`.  I think we wanted to avoid `malloc()` and the bigger one is intentional.

**Comment by steven-johnson** [comment]
> Good catch, yikes! I definitely did not understand this was the behavior.

The API is really a bit too clever for its own good

**Comment by slomp** [comment]
When I implemented `StackPrinter`, it was to avoid malloc since I needed larger buffers for tracing / error dumps. The class had it's own internal `char buffer [length]` that was passed to `Printer<type, length>` to "override" the internal `malloc()`.

What seems to have happened is that @abadams did some improvements on Halide's `Printer` and decided to retire my `StackPrinter` (this is the commit for reference 2e5309aed8cfc8fe3a9d182f561e806844b528de)

**Comment by slomp** [comment]
> Good catch, yikes! I definitely did not understand this was the behavior.
> 
> Pinging @slomp to chime in about the one in `d3d12compute.cpp`. I think we wanted to avoid `malloc()` and the bigger one is intentional.

Yup, malloc'ing on the 64k case is fine (it's on a debug path, just to dump shader code). Everywhere else, stack allocation is preferred, since it's small and scoped (very short lived).

**Comment by steven-johnson** [comment]
> Everywhere else, stack allocation is preferred, since it's small and scoped (very short lived).

Right -- and the point is that that only happens currently in those two cases (16 and 256)

**Comment by slomp** [comment]
Removing `StackPrinter` here: https://github.com/halide/Halide/pull/6470/files

As for the regular `Printer`, I say we just change the default template parameter from 1024` to `256`.

(Although, to be honest, I kind of prefer to also have an explicit `StackPrinter` available since that has a precise behavior).

**Comment by abadams** [comment]
IIRC I had to remove the large stack allocations because it was causing failures to even compile the runtime in some circumstances.

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — some fraction of these are expected to be wrong.

**Reason (CLOSE_FIXED):** The reported wart — a `scratch` array silently collapsing to 1 byte for any length over 256 — is gone. `printer.h` was restructured into `PrinterBase` plus an explicit `HeapPrinter` (always `malloc`) and an explicit `StackPrinter` (real stack buffer, `static_assert`ed to be \<= 256 bytes).

**Evidence:** Commit `59118dea3` "Deal with Printer::scratch (#6469) (#6472)"; current code at `src/runtime/printer.h:184`-`203` (`HeapPrinter`) and `src/runtime/printer.h:245`-`253` (`StackPrinter` with `static_assert(buffer_length <= 256, ...)`).

Fixed by #6472 (`59118dea3`): `Printer` was split into an explicit `HeapPrinter` (always mallocs) and an explicit `StackPrinter` (real stack buffer, `static_assert`ed to be \<= 256 bytes), so the silently-ignored `scratch` field no longer exists. Closing.
